src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,5392 @@
+	/* num_pages = length / page_size */
+	int              num_pages;
+	int              num_dirty_pages;
+	off_t            offset;
+	/* An array to indicate whether a page is updated or not */
+	bool            *updated;
+};
+
+struct fd_dup2 {
+	int fd_src, fd_dest;
+	bool dest_closed;
+};
+
+/* working dir of current process */
+static char            cur_dir[DFS_MAX_PATH] = "";
+static bool            segv_handler_inited;
+/* Old segv handler */
+struct sigaction       old_segv;
+
+/* the flag to indicate whether initlization is finished or not */
+static bool            hook_enabled;
+static pthread_mutex_t lock_dfs;
+static pthread_mutex_t lock_fd;
+static pthread_mutex_t lock_dirfd;
+static pthread_mutex_t lock_mmap;
+static pthread_mutex_t lock_fd_dup2ed;
+
+/* store ! umask to apply on mode when creating file to honor system umask */
+static mode_t          mode_not_umask;
+
+static void
+finalize_dfs(void);
+static void
+update_cwd(void);
+
+struct dir_hdl {
+	d_list_t   entry;
+	dfs_obj_t *oh;
+	char       name[DFS_MAX_PATH];
+};
+
+static inline struct dir_hdl *
+hdl_obj(d_list_t *rlink)
+{
+	return container_of(rlink, struct dir_hdl, entry);
+}
+
+static bool
+key_cmp(struct d_hash_table *htable, d_list_t *rlink, const void *key, unsigned int ksize)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return (strncmp(hdl->name, (const char *)key, DFS_MAX_PATH) == 0);
+}
+
+static void
+rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	int             rc;
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	rc = dfs_release(hdl->oh);
+	if (rc == ENOMEM)
+		rc = dfs_release(hdl->oh);
+	if (rc && bLog)
+		D_WARN("dfs_release() failed: %s", strerror(rc));
+	D_FREE(hdl);
+}
+
+static bool
+rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	return true;
+}
+
+static uint32_t
+rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return d_hash_string_u32(hdl->name, strnlen(hdl->name, DFS_MAX_PATH));
+}
+
+static d_hash_table_ops_t hdl_hash_ops = {.hop_key_cmp    = key_cmp,
+					  .hop_rec_decref = rec_decref,
+					  .hop_rec_free   = rec_free,
+					  .hop_rec_hash   = rec_hash};
+
+static int
+lookup_insert_dir(int idx_dfs, const char *name, mode_t *mode, dfs_obj_t **obj)
+{
+	struct dir_hdl *hdl;
+	dfs_obj_t      *oh;
+	d_list_t       *rlink;
+	size_t          len;
+	int             rc;
+
+	if (obj == NULL) {
+		D_ERROR("Invalid pointer!");
+		return -EFAULT;
+	}
+	*obj = NULL;
+	len = strnlen(name, DFS_MAX_PATH);
+	if (len >= DFS_MAX_PATH) {
+		D_ERROR("path is too long!");
+		return -ENAMETOOLONG;
+	}
+	rlink = d_hash_rec_find(dfs_list[idx_dfs].dfs_dir_hash, name, len);
+	if (rlink != NULL) {
+		hdl = hdl_obj(rlink);
+		*obj = hdl->oh;
+		return 0;
+	}
+
+	rc = dfs_lookup(dfs_list[idx_dfs].dfs, name, O_RDWR, &oh, mode, NULL);
+	if (rc)
+		return rc;
+
+	if (mode && !S_ISDIR(*mode)) {
+		*obj = oh;
+		return 0;
+	}
+
+	D_ALLOC_PTR(hdl);
+	if (hdl == NULL)
+		return -ENOMEM;
```

Comment:
You leak oh here.

Reply:
You are right. I will add "dfs_release(oh);" before return. Thank you very much!